### PR TITLE
[query] Aggregator package SCode omnibus commit

### PR DESCRIPTION
**Pull request by tpoterba**
This PR pushes aggregators much closer to where we want
them to be with regardes to SCodes/STypes. Here are the
important conceptual changes:
1. Aggregators are no longer parameterized by the ptypes
   of seqop and initop args. Instead, the state signature
   contains a sequence of VirtualTypeWithRequiredness, which
   is exactly what its name says. Aggregators use ptypes
   internally, but this is not in the state signature.
2. As a consequence of 1), we no longer cast argument types
   to seqop/initops. We are still able to, for instance,pass a seqop
   argument of a different type than the container type to
   the collect aggregator because the collect aggregator accepts
   an SCode argument, and uses the appropriate PType constructor to
   store that SCode (with a possible conversion if necessary, as here).
3. We codebuilder-ify most of the aggregator package. There are some
   straggling bits, but I will clean that up in a followup since the
   time to get a change in will scale super-linearly with its size.

Benchmarks forthcoming.

Agg SCodes hopefully done

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/Code.scala [hunk 1]
```diff
@@ -898,6 +898,8 @@ class CodeLong(val lhs: Code[Long]) extends AnyVal {
   def toD: Code[Double] = Code(lhs, lir.insn1(L2D))
 
   def toS: Code[String] = Code.invokeStatic1[java.lang.Long, Long, String]("toString", lhs)
+
+  def hexString: Code[String] = Code.invokeStatic1[java.lang.Long, Long, String]("toHexString", lhs)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/asm4s/Code.scala:902]: Used in debugging process, but leaving for future debugging.

File: hail/python/test/hail/experimental/test_codec.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import hail as hl
 from test.hail.helpers import *
 
+setUpModule = startTestHailContext
```
Comment (tpoterba) [inline hail/python/test/hail/experimental/test_codec.py:4]: missing from the file, added to let me run subsets of tests

File: hail/python/test/hail/expr/test_ndarrays.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 
 from hail.utils.java import FatalError, HailUserError
 
+setUpModule = startTestHailContext
+tearDownModule = stopTestHailContext
```
Comment (tpoterba) [inline hail/python/test/hail/expr/test_ndarrays.py:9]: missing from the file, added to let me run subsets of tests

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 4]
```diff
@@ -547,23 +547,22 @@ class Emit[C](
         assert(aggs(i) == sig.state)
         val rvAgg = agg.Extract.getAgg(sig)
 
-        val argVars = args.zip(rvAgg.initOpTypes).map { case (a, t) =>
-          emit(a, container = container.flatMap(_.nested(i, init = true)))
-            .castTo(mb, region.code, t)
-        }.toArray
+        val argVars = args
+          .map { a => emit(a, container = container.flatMap(_.nested(i, init = true))) }
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:551]: important: no more casts!

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 5]
```diff
@@ -547,23 +547,22 @@ class Emit[C](
...
         val AggContainer(aggs, sc, _) = container.get
         assert(sig.state == aggs(i))
         val rvAgg = agg.Extract.getAgg(sig)
 
-        val argVars = args.zip(rvAgg.seqOpTypes).map { case (a, t) =>
-          emit(a, container = container.flatMap(_.nested(i, init = false)))
-            .castTo(mb, region.code, t)
-        }.toArray
+        val argVars = args
+          .map { a => emit(a, container = container.flatMap(_.nested(i, init = false))) }
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:563]: important: no more casts!

File: hail/src/main/scala/is/hail/expr/ir/EmitCodeBuilder.scala [hunk 6]
```diff
@@ -137,4 +144,9 @@ class EmitCodeBuilder(val emb: EmitMethodBuilder[_], var code: Code[Unit]) exten
   def printRegionValue(value: Code[_], typ: PType, region: Value[Region]): Unit = {
     append(Code._println(StringFunctions.boxArg(EmitRegion(emb, region), typ)(value)))
   }
+
+  // for debugging
+  def strValue(r: Value[Region], t: PType, code: Code[_]): Code[String] = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/EmitCodeBuilder.scala:149]: used in debugging process

File: hail/src/main/scala/is/hail/expr/ir/EmitStream.scala [hunk 7]
```diff
@@ -1247,7 +1247,8 @@ object EmitStream {
                     keyRegion.clear(),
                     EmitCodeBuilder.scopedVoid(mb) { cb =>
                       val pc = new SSubsetStructCode(keyViewType, xCurElt.load().asBaseStruct)
-                      cb.assign(xCurKey, pc.castTo(cb, keyRegion.code, keyType))},
+                      cb.assign(xCurKey, pc.castTo(cb, keyRegion.code, keyType, deepCopy = true))
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/EmitStream.scala:1250]: latent bug in EmitStream

Comment (chrisvittal) [reply hail/src/main/scala/is/hail/expr/ir/EmitStream.scala:1250]: Do we know the cause?

**Comment by chrisvittal** [comment]
I'm ready to merge this. What's the perf status?

**Comment by tpoterba** [comment]
> I'm ready to merge this

Woohoo!

I'm in a state of deep depression about the replicability of benchmarks right now. They seem (subjectively) even more noisy from run to run than things were six months ago. This could be 1-2% slower overall, or identical to main -- I can't tell, since benchmarks of main vary by nearly 5% from day to day.

Either way, I'm good to merge this even with the worst-case 2% degradation. Progress toward stack struct is worth it.

**Review by chrisvittal — APPROVED** [review]

**Comment by tpoterba** [comment]
😱 

Thanks Chris!

**Pull request merged**